File: solution/40. Combination Sum II/40.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(), candidates.end());
        dfs(candidates, target, 0);
        return res;
    }
    
private:
    vector<vector<int>> res;
    vector<int> cur;
    
    void dfs(vector<int>& A, int target, int start) {
        if (target < 0) return;
        if (target == 0) {
            res.push_back(cur);
            return;
        }
        
        for (int i = start; i < A.size(); ++i) {
            if (start < i && A[i - 1] == A[i]) continue;
            cur.push_back(A[i]);
            dfs(A, target - A[i], i + 1);
            cur.pop_back();
        }
    }
};

File: solution/40. Combination Sum II/40.py
class Solution:
    def combinationSum2(self, A: List[int], target: int) -> List[List[int]]:
        A.sort()
        res = []
        
        def dfs(target, start, cur, res):
            if target < 0:
                return
            if target == 0:
                res.append(cur)
                return
            
            for i in range(start, len(A)):
                if start < i and A[i - 1] == A[i]:
                    continue
                dfs(target - A[i], i + 1, cur + [A[i]], res)
        
        dfs(target, 0, [], res)
        return res
    
